src/software/visualizer/drawing/navigator.cpp

```diff
@@ -18,19 +15,37 @@ AIDrawFunction drawNavigator(std::shared_ptr<Navigator> navigator)
         pen.setWidth(2);
         pen.setCosmetic(true);
 
-        QBrush brush(navigator_path_color);
-        brush.setStyle(Qt::BrushStyle::SolidPattern);
-
         for (const auto& path : planned_paths)
         {
             for (size_t i = 1; i < path.size(); i++)
             {
                 Segment path_segment(path[i - 1], path[i]);
-                QLineF line = createQLineF(path_segment);
-                scene->addLine(line, pen);
+                drawSegment(scene, path_segment, pen);
             }
         }
+
+        for (const auto& obstacle : obstacles)
+        {
+            drawObstacle(scene, obstacle->getObstacleShape(), pen);
+        }
     };
 
     return AIDrawFunction(draw_function);
 }
+
+void drawObstacle(QGraphicsScene* scene, const ObstacleShape& obstacle_shape,
+                  const QPen& pen)
+{
+    std::visit(
+        overload{[scene, &obstacle_shape, pen](const ConvexPolygon& convex_polygon) {
+                     drawConvexPolygon(scene, convex_polygon, pen);
+                 },
+                 [scene, &obstacle_shape, pen](const Circle& circle) {
+                     drawCircle(scene, circle, pen);
+                 },
+                 [](auto arg) {
```

Comment:
Does providing a function that matches on `auto` prevent us from getting compiler warning if we aren't handling a type explicitly? It might be like how we only get compiler warnings for `switch` statements only if we don't provide a `default` case. We definitely want the compiler to catch this so please double check.

Reply:
sadly no